staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/established_controller.go

```diff
@@ -0,0 +1,101 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"k8s.io/apimachinery/pkg/util/wait"
+	"time"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/staging/src/k8s.io/client-go/util/workqueue"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/internalclientset"
+	"github.com/golang/glog"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+// EstablishingController controls how and when CRD is established.
+type EstablishingController struct {
+	client *internalclientset.Clientset
+	queue workqueue.DelayingInterface
+}
+
+// NewEstablishingController creates new EstablishingController.
+func NewEstablishingController(client *internalclientset.Clientset) *EstablishingController {
+	return &EstablishingController{
+		client: client,
+		queue: workqueue.NewDelayingQueue(),
+	}
+}
+
+func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	glog.Infof("Starting NamingConditionController")
+	defer glog.Infof("Shutting down NamingConditionController")
+
+	// only start one worker thread since its a slow moving API and the naming conflict resolution bits aren't thread-safe
+	go wait.Until(ec.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (ec *EstablishingController) runWorker() {
+	for ec.establishCRD() {
+	}
+}
+
+// EstablishCRD turns CRD into established and returns
+// is it successful or error
+func (ec *EstablishingController) establishCRD() bool {
+	// We will get back to this for{} :)
+		key, quit := ec.queue.Get()
+		if quit {
+			return false
+		}
+		defer ec.queue.Done(key)
+		crd, _ := ec.client.Apiextensions().CustomResourceDefinitions().Get(key.(string), v1.GetOptions{})
+
+		// Poll because if multiple servers try to change CRD (e.g. in HA) this can fail.
+		// That failure is normal, and so, we're trying to set it to Established again.
+		err := wait.Poll(500 * time.Millisecond, 3 * time.Second, func() (bool, error) {
+			if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) &&
+				apiextensions.IsCRDConditionTrue(crd, apiextensions.Installing) {
+				establishedCondition := apiextensions.CustomResourceDefinitionCondition{
+					Type:    apiextensions.Established,
+					Status:  apiextensions.ConditionTrue,
+					Reason:  "InitialNamesAccepted",
+					Message: "the initial names have been accepted",
+				}
+
+				// Update storage with new CRD condition.
+				apiextensions.SetCRDCondition(crd, establishedCondition)
+
+				// Update server with new CRD condition.
+				var err error
+				_, err = ec.client.Apiextensions().CustomResourceDefinitions().UpdateStatus(crd)
+				if err != nil {
+					return false, err
```

Comment:
this is not enough: imagine another server updates the crd meanwhile. Then the CRD's `meta.ResourceVersion` of your instance here will have an old value. The apiserver will reply with a conflict error for that that reason (failed optimistic concurrency). So you have to get the latest version of the CRD and try again with your update.

In other words: the current code (the wait.Poll) will always use the old crd instance. If the update fails with a conflict, the current code will fail again and again (until 3 seconds are up).

If you just stop the whole updating attemp instead, the update of the CRD will arrive at the informer eventually and your controller will get another chance to update. So I guess we don't even have to requeue. All is automatic. Compare the naming controller if that uses this logic as well.

Reply:
That totally makes sense. Thank for pointing this out.

The requeue is not needed at this point, but I've added it later. I've updated this code to get new CRD in every Poll's iteration/try. That should be better I think.

The questionable part is: if `establishCRD()` fails, should I requeue the key and after how long? In this review of EstablishingController it requeues it after the failure in the `proccessNextWorkItem` function.